### PR TITLE
logging: add logging mechanism, hooks to log uncaught crashes, and an example use of the logging tools

### PR DESCRIPTION
**Pull request by ianohara**
This adds logging tools in the `octopi.logging` module (along with the `octopi` package), and two helpers there for getting loggers and setting the `octopi` logger logging level.  This allows us to have a dedicated hierarchy of loggers that we can configure while leaving all the loggers setup by other libraries alone.

It also adds an example usage in the `control._def.py` module.  We can start converting individual files after this.

It also lays the groundwork for better crash reports by allowing us to hook up handlers for uncaught exceptions.  By default these just print to console, but we want to be able to hook up a `FileHandler` to our logger and have the uncaught exceptions go there as well.  Actually creating the `FileHandler` and attaching it to the root octopi logger is to come.

Tested by:
Attempting to run `main_hcs.py`, and seeing log messages:
image

I also added a random `raise ValueError("test test")` in `main_hcs.py` for a different run to make sure that the handlers work as expected:
image

### CODE REVIEW

**Comment by ianohara** [comment]
I will rebase on #113 once (if) it lands.  I can't select a pull request base that's not pushed to this fork, so can't get rid of that noise.  Sorry :/

File: software/octopi/logging.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+
+    logger = get_logger()
+
+    def new_excepthook(exception_type: Type[BaseException], value: BaseException, tb: TracebackType):
+        try:
+            handler(exception_type, value, tb)
+        except BaseException as e:
+            logger.critical("Custom excepthook handler raised exception", e)
+        if call_existing_too:
+            old_excepthook(exception_type, value, tb)
```
Comment (ianohara) [inline software/octopi/logging.py:103]: I should rename the tb here and in `new_excepthook` to `tb_type`.  The other two hooks actually get tracebacks (so the `type(tb)` is needed).

**Comment by hongquanli** [comment]
> This adds logging tools in the `octopi.logging` module (along with the `octopi` package), and two helpers there for getting loggers and setting the `octopi` logger logging level. This allows us to have a dedicated hierarchy of loggers that we can configure while leaving all the loggers setup by other libraries alone.
> 
> It also adds an example usage in the `control._def.py` module. We can start converting individual files after this.
> 
> It also lays the groundwork for better crash reports by allowing us to hook up handlers for uncaught exceptions. By default these just print to console, but we want to be able to hook up a `FileHandler` to our logger and have the uncaught exceptions go there as well. Actually creating the `FileHandler` and attaching it to the root octopi logger is to come.
> 
> Tested by: Attempting to run `main_hcs.py`, and seeing log messages: image
> 
> I also added a random `raise ValueError("test test")` in `main_hcs.py` for a different run to make sure that the handlers work as expected: image

@ianohara can you change octopi to squid?

**Comment by ianohara** [comment]
> > This adds logging tools in the `octopi.logging` module (along with the `octopi` package), and two helpers there for getting loggers and setting the `octopi` logger logging level. This allows us to have a dedicated hierarchy of loggers that we can configure while leaving all the loggers setup by other libraries alone.
> > It also adds an example usage in the `control._def.py` module. We can start converting individual files after this.
> > It also lays the groundwork for better crash reports by allowing us to hook up handlers for uncaught exceptions. By default these just print to console, but we want to be able to hook up a `FileHandler` to our logger and have the uncaught exceptions go there as well. Actually creating the `FileHandler` and attaching it to the root octopi logger is to come.
> > Tested by: Attempting to run `main_hcs.py`, and seeing log messages: image
> > I also added a random `raise ValueError("test test")` in `main_hcs.py` for a different run to make sure that the handlers work as expected: image
> 
> @ianohara can you change `octopi` to `squid`?

Yep, will do.  Also I'll move this over to https://github.com/Cephla-Lab/Squid

**Pull request closed without merging**